hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/stack/SavepointStackImpl.java

```diff
@@ -156,4 +225,205 @@ public WritableStates getWritableStates(@NonNull final String serviceName) {
+        requireNonNull(builderClass);
+        requireNonNull(consumer);
+        if (builderSink != null) {
+            builderSink.forEachOtherBuilder(consumer, builderClass, baseBuilder);
+        }
+        for (var savepoint : stack) {
+            savepoint.forEachOtherBuilder(consumer, builderClass, baseBuilder);
+        }
+    }
+
+    /**
+     * Returns the {@link HandleContext.TransactionCategory} of the transaction that created this stack.
+     * @return the transaction category
+     */
+    public HandleContext.TransactionCategory txnCategory() {
+        return baseBuilder.category();
+    }
+
+    /**
+     * Creates a new stream builder for a removable child in the active savepoint.
+     * @return the new stream builder
+     */
+    public SingleTransactionRecordBuilder createRemovableChildBuilder() {
+        return peek().createBuilder(REMOVABLE, CHILD, NOOP_RECORD_CUSTOMIZER, false);
+    }
+
+    /**
+     * Creates a new stream builder for a reversible child in the active savepoint.
+     * @return the new stream builder
+     */
+    public SingleTransactionRecordBuilder createReversibleChildBuilder() {
+        return peek().createBuilder(REVERSIBLE, CHILD, NOOP_RECORD_CUSTOMIZER, false);
+    }
+
+    /**
+     * Creates a new stream builder for an irreversible preceding transaction in the active savepoint.
+     * @return the new stream builder
+     */
+    public SingleTransactionRecordBuilder createIrreversiblePrecedingBuilder() {
+        return peek().createBuilder(IRREVERSIBLE, PRECEDING, NOOP_RECORD_CUSTOMIZER, false);
+    }
+
+    /**
+     * Returns all following child records in the stack for use in end-of-EVM-transaction throttling.
+     * <p>
+     * To be removed on completion of HIP-993 and adoption of per-dispatch throttling.
+     *
+     * @return the list of child records
+     */
+    @Deprecated
+    public List<SingleTransactionRecordBuilder> getChildBuilders() {
+        final var childRecords = new ArrayList<SingleTransactionRecordBuilder>();
+        for (final var savepoint : stack) {
+            for (final var builder : savepoint.followingBuilders()) {
+                if (builder.category() == CHILD) {
+                    childRecords.add(builder);
+                }
+            }
+        }
+        return childRecords;
+    }
+
+    /**
+     * Returns the top savepoint without removing it from the stack. Used only by the {@link WritableStatesStack},
+     * not part of the public API.
+     *
+     * @return the top savepoint
+     * @throws IllegalStateException if the stack has been committed already
+     */
+    @NonNull
+    Savepoint peek() {
+        if (stack.isEmpty()) {
+            throw new IllegalStateException("The stack has already been committed");
+        }
+        return stack.peek();
+    }
+
+    /**
+     * Builds and caches the result of the user transaction with
```

Comment:
Where does this method cache the result? I don't see any caching here

Reply:
I just meant the return of this method is used in `recordCache`. Changed the description

Reply:
Ah got it, thanks!